torch/_higher_order_ops/associative_scan.py

```diff
@@ -401,9 +443,421 @@ def associative_scan_op_dense(combine_fn, xs, additional_inputs):
     return generic_associative_scan(combine_fn, xs, additional_inputs=additional_inputs)
 
 
-associative_scan_op.py_impl(DispatchKey.Autograd)(
-    autograd_not_implemented(associative_scan_op, deferred_error=True)
-)
+class AssociativeScanAutogradOp(torch.autograd.Function):
+    r"""
+    Example::
+        xs = torch.arange(1, 5) = [1, 2, 3, 4]
+        ys = torch.cumprod(xs) = [1, 2, 6, 24]
+
+        def combine_fn(a: torch.Tensor, b: torch.Tensor):
+            return a * b
+
+        The ``combine_fn_bw``, computing the gradients for a and b of ``combine_fn`` is computed as:
+        def combine_fn_bw(a: torch.Tensor, b: torch.Tensor, g_y: torch.Tensor):
+            return g_y * b, g_y * a
+
+        Note: In a real usecase of associative_scan, there may be additional_inputs that participate in the
+        forward as well as in the backward of the scan operator. For the sake of readability those inputs
+        have been omitted in the following example, but are included in the subsequent detailed description below
+
+        The forward output of associative_scan is computed as:
+        ys = associative_scan(combine_fn, xs).
+
+        For example, this computation can be unpacked as:
+        ys_0 = xs_0
+        ys_1 = combine_fn(ys_0, xs_1)
+        ...
+        ys_T = combine_fn(ys_(T-1), xs_T)
+
+        Note: In a real usecase of associative_scan this operation is parallelized from O(T) to O(log(T)).
+
+        Given the ys, the gradients for xs can be computed as follows:
+        We receive the upstream gradients in torch.autograd.Function, i.e., we get g_ys,
+        where g_ys is the vector of all intermediate gradients of the outputs [g_ys_0, g_ys_1, ..., g_ys_T]
+
+        We then proceed to compute the gradients for the xs (g_xs) by computing for every step:
+        the instantaneous for g_x_t, as well as
+        the gradient for the previous intermediate results g_h_t using
+        g_h_t, g_x_t = ctx._combine_fn_bw(ys_(t-1), xs_t)
+
+        For example:
+        g_h_0, g_x_0 = [1, 1]
```

Comment:
I'm confused how we get these two, i thought we have xs, ys (saved by forward), and g_ys (backward inputs)?